openshift-kube-apiserver/admission/customresourcevalidation/authentication/validate_authentication.go

```diff
@@ -121,14 +151,184 @@ func validateAuthenticationSpec(spec configv1.AuthenticationSpec) field.ErrorLis
+	errs := field.ErrorList{}
+	warns := []string{}
+	for i, extra := range extras {
+		err, warn := validateExtra(ctx, path.Child("extra").Index(i), cel, extra)
+		errs = append(errs, err...)
+		if len(warn) > 0 {
+			warns = append(warns, warn)
+		}
+	}
+	return errs, warns
+}
+
+func validateExtra(ctx context.Context, path *field.Path, cel *celStore, extra configv1.ExtraMapping) (field.ErrorList, string) {
+	childPath := path.Child("valueExpression")
+	err, warn := validateCELExpression(ctx, cel, &authenticationcel.ExtraMappingExpression{
+		Key:        extra.Key,
+		Expression: extra.ValueExpression,
+	})
+
+	if len(warn) > 0 {
+		warn = fmt.Sprintf("validating %s: %s", childPath, warn)
+	}
+
+	if err != nil {
+		return field.ErrorList{field.Invalid(childPath, extra.ValueExpression, err.Error())}, warn
+	}
+
+	return nil, warn
+}
+
+type compileResult struct {
+	err  error
+	warn string
+}
+
+func validateCELExpression(ctx context.Context, cel *celStore, accessor authenticationcel.ExpressionAccessor) (error, string) {
+	// if context has been canceled, don't try to compile any expressions
+	if err := ctx.Err(); err != nil {
+		return err, ""
+	}
+
+	// expression is currently being compiled
+	if val, ok := cel.compilingStore.Get(accessor.GetExpression()); ok {
+		// we use a channel to determine if it has finished compiling
+		select {
+		case res := <-val.(chan compileResult):
+			// done compiling
+			return res.err, res.warn
+		case <-ctx.Done(): // handle the case of context cancelling while waiting
+			return ctx.Err(), ""
+		}
+	}
+
+	// if the expression is not currently being compiled, it might have already been compiled
+	// either successfully or unsucessfully
+	if val, ok := cel.compiledStore.Get(accessor.GetExpression()); ok {
+		if val != nil {
+			res := val.(compileResult)
+			return res.err, res.warn
+		}
+
+		return nil, ""
+	}
+
+	// expression is not currently being compiled, and has not been compiled before (or has been long enough since it was last compiled that we dropped it).
+	// Let's compile it.
+	warningString := ""
+	compiled := make(chan error)
+	defer close(compiled)
+
+	compiling := make(chan compileResult, 1)
+	defer close(compiling)
+
+	cel.compilingStore.Add(accessor.GetExpression(), compiling)
+	defer cel.compilingStore.Remove(accessor.GetExpression())
+
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				// convert the panic into an error state for the expression
+				compiled <- fmt.Errorf("recovered from a panic while compiling expression %q: %v", accessor.GetExpression(), r)
+			}
+		}()
+
+		_, err := cel.compiler.CompileClaimsExpression(accessor)
+
+		compiled <- err
+	}()
+
+	timer := time.AfterFunc(time.Second, func() {
+		warningString = fmt.Sprintf("cel expression %q took more than 1 second to compile", accessor.GetExpression())
+	})
+
+	err := <-compiled
+
+	timer.Stop()
+
+	compilationResult := compileResult{
+		err, warningString,
+	}
```

Comment:
Your anonymous function will run in a separate goroutine, so you have a potential data race here (between the write on line 318 and the read on line 326). Check out the instructions on [Stop](https://pkg.go.dev/time#Timer.Stop):

> For a func-based timer created with [AfterFunc](https://pkg.go.dev/time#AfterFunc)(d, f), if t.Stop returns false, then the timer has already expired and the function f has been started in its own goroutine; Stop does not wait for f to complete before returning. If the caller needs to know whether f is completed, it must coordinate with f explicitly.